components/script/dom/bindings/refcounted.rs

```diff
@@ -153,59 +132,55 @@ impl LiveDOMReferences {
-                refcount.clone()
-            }
+            Occupied(mut entry) => match entry.get().upgrade() {
+                Some(refcount) => refcount,
+                None => {
+                    let refcount = Arc::new(TrustedReference::new(ptr));
+                    entry.insert(Arc::downgrade(&refcount));
+                    refcount
+                },
+            },
             Vacant(entry) => {
-                let refcount = Arc::new(Mutex::new(1));
-                entry.insert(refcount.clone());
+                let refcount = Arc::new(TrustedReference::new(ptr));
+                entry.insert(Arc::downgrade(&refcount));
                 refcount
             }
         }
     }
+}
 
-    /// Unpin the given DOM object if its refcount is 0.
-    pub fn cleanup(raw_reflectable: TrustedReference) {
-        let TrustedReference(raw_reflectable) = raw_reflectable;
-        LIVE_REFERENCES.with(|ref r| {
-            let r = r.borrow();
-            let live_references = r.as_ref().unwrap();
-            let mut table = live_references.table.borrow_mut();
-            match table.entry(raw_reflectable) {
-                Occupied(entry) => {
-                    if *entry.get().lock().unwrap() != 0 {
-                        // there could have been a new reference taken since
-                        // this message was dispatched.
-                        return;
-                    }
-
-                    let _ = entry.remove();
-                }
-                Vacant(_) => {
-                    // there could be a cleanup message dispatched, then a new
-                    // pinned reference obtained and released before the message
-                    // is processed, at which point there would be no matching
-                    // hashtable entry.
-                    info!("attempt to cleanup an unrecognized reflector");
-                }
-            }
-        })
+/// Remove null entries from the live references table
+fn remove_nulls<K: Eq + Hash + Clone, V> (table: &mut HashMap<K, Weak<V>>) {
+    let to_remove: Vec<K> =
+        table.iter()
+        .filter(|&(_, value)| Weak::upgrade(value).is_none())
+        .map(|(key, _)| key.clone())
+        .collect();
+    info!("removing {} refcounted references", to_remove.len());
+    for key in to_remove {
+        table.remove(&key);
     }
 }
 
 /// A JSTraceDataOp for tracing reflectors held in LIVE_REFERENCES
 pub unsafe extern "C" fn trace_refcounted_objects(tracer: *mut JSTracer,
                                                   _data: *mut os::raw::c_void) {
-    debug!("tracing live refcounted references");
+    info!("tracing live refcounted references");
     LIVE_REFERENCES.with(|ref r| {
         let r = r.borrow();
         let live_references = r.as_ref().unwrap();
-        let table = live_references.table.borrow();
+        let mut table = live_references.table.borrow_mut();
+        remove_nulls(&mut table);
```

Comment:
I didn't say we could use it for `Trusted`, I was just saying we have a precedent for cleaning stuff up on GC trace.

As for your question, `WeakRefVec` doesn't implement `JSTraceable` itself because cleaning up needs mutable access to it.